AgreeToTerms/Features/PolicyAgreement/ViewModel/PolicyAgreementViewModel.swift

```diff
@@ -0,0 +1,84 @@
+//
+//  PolicyAgreementViewModel.swift
+//  AgreeToTerms
+//
+//  Created by Ran Helfer on 25/07/2022.
+//
+
+import UIKit
+import Combine
+
+/*
+ COMMENTS:
+ - created a protocol for the public view model API
+ - instead of exposing a `CurrentValueSubject` (which lets anyone publish events), use a Publisher, which only lets observing
+ - instead of returning a `WKNavigationActionPolicy` just return a `bool`
+ */
+
+protocol PolicyAgreementViewModelProtocol {
+    var canMoveToNextPage: AnyPublisher<Bool, Never> { get }
+    var currentPage: AnyPublisher<Int, Never> { get }
+    var policySignEnded: AnyPublisher<Bool, Never> { get }
+
+    func currentURL() -> String
+    func approveButtonPressed()
+    func shouldAllowPage(absoluteString: String?) -> Bool
+    func updateModelWithScroll(contentSizeHeight: CGFloat, scrollViewHeightBound: CGFloat, contentOffsetY: CGFloat)
+}
+
+final class PolicyAgreementViewModel: PolicyAgreementViewModelProtocol {
+    private let urlForPage = ["https://www.viz.ai/eula",
+                              "https://www.viz.ai/privacy-policy"]
+
+    let canMoveToNextPage: AnyPublisher<Bool, Never>
+    let currentPage: AnyPublisher<Int, Never>
+    let policySignEnded: AnyPublisher<Bool, Never>
+
+    private let canMoveToNextPageSubject = CurrentValueSubject<Bool, Never>(false)
+    private let currentPageSubject = CurrentValueSubject<Int, Never>(0)
+    private let policySignEndedSubject = CurrentValueSubject<Bool, Never>(false)
+
+    init() {
+        self.canMoveToNextPage = canMoveToNextPageSubject.eraseToAnyPublisher()
+        self.currentPage = currentPageSubject.eraseToAnyPublisher()
+        self.policySignEnded = policySignEndedSubject.eraseToAnyPublisher()
+    }
+
```

Comment:
@mschuetz-viz - i think i have a better solution rather than have both `CurrentValueSubject` variables and `AnyPublisher` variables. 

We can have just regular private types on our model - Bool, Int, String etc. And whenever we set a value we can assemble a simple `Output` struct that the controller has access to all the assembled published values. 

Personally, i think it is much better - see bellow

```


protocol PolicyAgreementViewModelProtocol {
    func currentURL() -> String
    func approveButtonPressed()
    func shouldAllowPage(absoluteString: String?) -> Bool
    func updateModelWithScroll(contentSizeHeight: CGFloat, scrollViewHeightBound: CGFloat, contentOffsetY: CGFloat)
    
    var numberOfPages: Int { get }
    var subjects: CurrentValueSubject<PolicyApproveValueSubject, Never>  { get }
}

struct PolicyApproveValueSubject {
    var canMoveToNextPage = false
    var currentPage = 0
    var policySignEnded = false
    
    init() {
        self.currentPage = 0
        self.canMoveToNextPage = false
        self.policySignEnded = false
    }
}

final class PolicyApproveModel: PolicyAgreementViewModelProtocol  {
    private let urlForPage = ["https://www.viz.ai/eula",
                              "https://www.viz.ai/privacy-policy"]
    var subjects = CurrentValueSubject<PolicyApproveValueSubject, Never>(PolicyApproveValueSubject())

    private var canMoveToNextPage: Bool = false {
        didSet {
            self.subjects.value.canMoveToNextPage = canMoveToNextPage
        }
    }
    
    private var currentPage: Int = 0 {
        didSet {
            self.subjects.value.currentPage = currentPage
        }
    }
    
    private var policySignEnded: Bool = false {
        didSet {
            self.subjects.value.policySignEnded = policySignEnded
        }
    }

```

Reply:
Some thoughts here:
1) Personally, I wouldn't opt for "packing" attributes into a struct just because of having less code. I'm not sure if `canMoveToNextPage`, `currentPage` and `policySignEnded` are enough related to justify wrapping them into one struct. For me they seem to be part of the view model, not any other, sub-related entity of the view model.

2) You still expose a `CurrentValueSubject` to the outside world, so that any client of `PolicyApproveModel` can modify the subject, which should be avoided. Example. the view controller still could use `viewModel.subjects.send()` to modify the subjects state.

3) The view model's properties `canMoveToNextPage`, `currentPage ` and `policySignEnded` can provoke state inconsistencies as they store state in parallel to `subjects`. Take for example the following code snippet:

```swift
    func test() {
        self.policySignEnded = true
        print("\(self.policySignEnded) - \(subjects.value.policySignEnded)") // true - true
        self.subjects.value.policySignEnded = false
        print("\(self.policySignEnded) - \(subjects.value.policySignEnded)") // true - false
    }
```

what you probably wanted to do is 

```swift
    private var policySignEnded: Bool {
        get { return self.subjects.value.policySignEnded }
        set { self.subjects.value.policySignEnded = newValue }
    }
```

which resolves the inconsistencies because get/ set is directly accessing `subjects`. Still, I don't see any advantage of having these properties, as they don't prevent the `subjects` being modified from the outside.

What you could do - if you think that 1) isn't a problem - is removing the view model's properties `canMoveToNextPage`, `currentPage ` and `policySignEnded`, and doing the `Subject` / `Publisher` dance:

```swift 
    private let dataSubject: CurrentValueSubject<PolicyApproveValueSubject, Never>
    let dataPublisher: AnyPublisher<PolicyApproveValueSubject, Never>

    init() {
        dataSubject = CurrentValueSubject<PolicyApproveValueSubject, Never>(PolicyApproveValueSubject())
        dataPublisher = dataSubject.eraseToAnyPublisher()
    }
```

Reply:
I am not sure at all that the fact that `CurrentValueSubject` can be abused by the view controller is a good enough reason to have all these duplications for any variable in the view model.  Abusing the model can be done in other ways as well and our practice can be simply not to do it in the `viewController`

I will definitely continue to think about this and probably huddle you on slack - good discussion thanks 🙏🏻